### ISSUE TITLE
ParamSpec: Allow only `P.args`

### ISSUE THREAD

**Issue by cdce8p** [python/typing#1000]
I would like to use `ParamSpec` to type a function that accepts a `Callable[P, Any]` and `*args` which will be passed to the callable. The function itself does not accept keyword arguments. As an example, `AbstractEventLoop.call_soon` behaves similar. This is the current Typeshed definition
```py
    def call_soon(self, callback: Callable[..., Any], *args: Any) -> Handle: ...
```
[typeshed -> stdlib/asyncio/events.pyi -> AbstractEventLoop](https://github.com/python/typeshed/blob/03861b5a1d317b14f516771f296e0325a13ed314/stdlib/asyncio/events.pyi#L90-L95)

```py
P = ParamSpec("P")

    def call_soon(self, callback: Callable[P, Any], *args: P.args) -> Handle: ...
```
Intuitively adding a ParamSpec variable like this would make sense: "Only except `*args`. If any arguments are passed, they need to match those of the `callback`." I would also expect that if `callback` has required arguments, they need to be passed with `*args` to `call_soon`.

The issue here is that PEP 612 explicitly forbids specifying `P.args` or `P.kwargs` alone. They always need to be together.
https://www.python.org/dev/peps/pep-0612/#id2

```
Furthermore, because the default kind of parameter in Python ((x: int)) may be addressed
both positionally and through its name, two valid invocations of a (*args: P.args, **kwargs: P.kwargs)
function may give different partitions of the same set of parameters. Therefore, we need to make
sure that these special types are only brought into the world together, and are used together,
so that our usage is valid for all possible partitions.
```

I do wonder if this strict limitation makes sense or if there are other ways to work around it so the case described above could be supported. After all it's similar to adding `**kwargs` to the function signature without ever passing keyword arguments.
```py
    def call_soon(self, callback: Callable[P, Any], *args: P.args, **kwargs: P.kwargs) -> Handle: ...
```

**Comment by TeamSpen210** [comment]
Perhaps this could require having `ParamSpec("Name", positional_only=True)`? Then it would be clear you don't care about keyword arguments, and any callable with required keyword args would be invalid.

**Comment by JelleZijlstra** [comment]
I believe PEP 646 would allow this: https://www.python.org/dev/peps/pep-0646/#type-variable-tuples-with-callable.

**Comment by erictraut** [comment]
@JelleZijlstra is correct. This is supported by PEP 646.

```python
Ts = TypeVarTuple("Ts")

def call_soon(self, callback: Callable[[*Ts], Any], *args: *Ts) -> Handle:
    ...
```

**Comment by cdce8p** [comment]
> This is supported by PEP 646.

Thanks for the responses! This certainly looks promising. I do wonder though why it wouldn't work with ParamSpec, too. The PEP says it's forbidden, but is there an actual limitation? I would have assumed that only specifying one would just set an additional constraint for the other to be empty. E.g. if only `P.args` is given, no kwargs should be accepted.

There is also the question about `kwargs` only calls. Something like this here seems reasonable to require keyword arguments for external functions.
```py
P = ParamSpec("P")
R = TypeVar("R")

def decorator(
    func: Callable[P, R]
) -> Callable[P, R]:
    def wrapper(**kwargs: P.kwargs) -> R:
        return func(**kwargs)
    return wrapper
```

**Comment by srittau** [comment]
I have a similar use case with [Ariadne middleware](https://ariadnegraphql.org/docs/middleware) functions, but with `kwargs`. In Ariadne, a "resolver" function takes two positional arguments and a varying number of keyword arguments that depend on the API the resolver implements. As the name implies, a middleware function sits between Ariadne and the resolver function and forwards arguments.

Ideally, I would like to type a middleware function as follows:

```python
_P = ParamSpec("_P")
_R = TypeVar("_R")
_T = TypeVar("_T")

def foo_middleware(
    resolver: Callable[Concatenate[_T, GraphQLResolveInfo, _P], _R],
    obj: _T,
    info: GraphQLResolveInfo,
    /,
    **kwargs: _P.kwargs,
) -> _R:
    ...
```

But currently this isn't possible, because `_P.args` is "missing".

**Comment by JelleZijlstra** [comment]
Closing as PEP 646 provides support for @cdce8p's use case.

**Comment by AlexWaygood** [comment]
> @erictraut mentioned in [microsoft/pyright#3775](https://github.com/microsoft/pyright/issues/3775) that this was be design. If so, `TypeVarTuple` can't be used to annotated functions like these.
> 
> @AlexWaygood mentioned a few other examples for TypeVarTuple in the typeshed tracking issue which wouldn't work then: [python/typeshed#8708 (comment)](https://github.com/python/typeshed/issues/8708#issuecomment-1241844607)

I think it would be unfortunate if this interpretation of the spec meant we had to stick with the (significantly more unsafe) signatures indefinitely for the many functions in `asyncio` that follow this pattern. I had been hoping we were close to being able to significantly increase the type safety of `asyncio`'s functions: see https://github.com/python/typeshed/pull/11015.

**Comment by erictraut** [comment]
If mypy figured out a way to implement this, I'll take another look at adding support in pyright. I'm not sure how mypy is capturing the value of `Ts` in the example above when `g` is passed as an argument. The captured type could be either `tuple[int]` or `tuple[int, str]`. Pyright's implementation chooses `tuple[int, str]`. It looks like mypy makes the same choice by default.

```python
def g(a: int, b: str = "") -> None: ...
def func(func: Callable[[*Ts], None]) -> tuple[*Ts]: ...
reveal_type(func(g))  # mypy and pyright both reveal "tuple[int, str]"
```

@JelleZijlstra, do you happen to know who implemented this in mypy? Was it Ivan? I could use some advice on how this works.

**Comment by JelleZijlstra** [comment]
Yes, Ivan did most of the PEP 646 work, though Jared Hance worked on it earlier.

**Comment by erictraut** [comment]
Good news — I figured out how to [make this work in pyright](https://github.com/microsoft/pyright/issues/6613) without too much hackery. This support will be included in the next release of pyright.

It's not clear from a reading of PEP 646 that this case is meant to be supported — and if so, what the intended behavior is. If typeshed is going to depend on this behavior, we should work to mandate this behavior in the typing spec and include it in the typing conformance test suite. I've added this to a long list that I've been compiling.
